extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufReader.java

```diff
@@ -80,7 +80,41 @@ public OpenCensusProtobufReader(
   @Override
   public CloseableIterator<InputRow> read()
   {
-    return CloseableIterators.withEmptyBaggage(readAsList().iterator());
+    return new CloseableIterator<InputRow>()
+    {
+      private boolean closed;
+      private Iterator<InputRow> metricIterator = null;
+
+      @Override
+      public boolean hasNext() throws ParseException
+      {
+        if (closed) {
+          return false;
+        }
+        if (metricIterator == null) {
+          metricIterator = readAsList().iterator();
+        }
+
+        return metricIterator.hasNext();
+      }
+
+      @Override
+      public InputRow next() throws ParseException
+      {
+        if (metricIterator == null) {
+          metricIterator = readAsList().iterator();
+        }
+        return metricIterator.next();
+      }
+
+      @Override
+      public void close()
+      {
+        if (!closed) {
+          closed = true;
+        }
+      }
+    };
```

Comment:
This can be done with less code using Guava's [Suppliers.memoize()](https://guava.dev/releases/19.0/api/docs/com/google/common/base/Suppliers.html#memoize(com.google.common.base.Supplier)).

Also, we can still use `CloseableIterators.withEmptyBaggage()` to reuse the `close` handling.

```suggestion
    Supplier<Iterator<InputRow>> supplier = Suppliers.memoize(() -> readAsList().iterator());
    return CloseableIterators.withEmptyBaggage(new Iterator<InputRow>() {
      @Override
      public boolean hasNext() {
        return supplier.get().hasNext();
      }

      @Override
      public InputRow next() {
        return supplier.get().next();
      }
    });
```

Reply:
Thank you Eric!